File: src/lib/controllers/Folder.ts
import { invoke } from "@tauri-apps/api";
import { reloadRepository } from "./Repository";
import type { TreeItem } from "$lib/components/structs/Tree";

export async function createFolder(name: string, parent: TreeItem) {
    return invoke('create_folder', {name, parent})
        .then((folder) => {
            reloadRepository();
            return Promise.resolve(folder as TreeItem);
        })
        .catch((err) => {
            return Promise.reject(err);
        })
}

export async function readFolder(folder: TreeItem) : Promise<TreeItem> {
    return invoke('read_folder', {folder: folder});
}

export async function updateFolder(origin: string, destination: string) : Promise<TreeItem> {
    return invoke('update_folder', {origin: origin, destination: destination});
}

export async function deleteFolder(folder: TreeItem) : Promise<TreeItem> {
    return invoke('delete_folder', {path: folder.path});
}

File: src-tauri/src/handlers/folder.rs
use std::{path::PathBuf, sync::Mutex};
use git2::Repository;
use tauri::{command, State};

use crate::core::{error::OpenDoorsError, folder::Folder, tree::TreeItem};

#[command] 
pub fn create_folder(state: State<'_, Mutex<Option<Repository>>>, name: String, parent: TreeItem) -> Result<TreeItem, OpenDoorsError> {
	let repo = state.lock().unwrap();
	Ok(Folder::create_folder(&repo, &name, &parent)?)
}

#[command]
pub fn read_folder(folder: TreeItem) -> Result<TreeItem, OpenDoorsError> {
	Ok(Folder::read(&folder)?)
}

#[command] 
pub fn update_folder(state: State<'_, Mutex<Option<Repository>>>, origin: PathBuf, destination: PathBuf) -> Result<TreeItem, OpenDoorsError> {
	let repo = state.lock().unwrap();
	Folder::update(&repo, &origin, &destination)?;
	Ok(TreeItem::from_path(&destination)?)
}

#[command] 
pub fn delete_folder(state: State<'_, Mutex<Option<Repository>>>, path: PathBuf) -> Result<(), OpenDoorsError> {
	let repo = state.lock().unwrap();
	Ok(Folder::delete(&repo, &path)?)
}
